## Installation
In a Jupyter notebook, install with the command

In [2]:
!pip3 install -U ucimlrepo 

In [13]:
from ucimlrepo import fetch_ucirepo, list_available_datasets
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 80)
import pprint

## List Available Datasets
### **Print a list of datasets that can be imported via `fetch_ucirepo`**

In [4]:
#list_available_datasets()
#list_available_datasets(filter='aim-ahead')   # only list datasets for AIM-AHEAD project
list_available_datasets(search='diabe')

--------------------------------------------------------------
The following datasets are available for search query "diabe":
--------------------------------------------------------------
Dataset Name                                     ID    
------------                                     --    
Diabetes 130-US Hospitals for Years 1999-2008    296   
Diabetic Retinopathy Debrecen                    329   
Early Stage Diabetes Risk Prediction             529   
CDC Diabetes Health Indicators                   891   



###Fetch Data from Diabetes 130-US Hospitals for Years 1999-2008    ID: 296 

In [5]:
diabetes = fetch_ucirepo(id=296)

## diabetes Object Quick Run Through

fetch_ucirepo(id) created a dictionary named diabetes 

In [6]:
type(diabetes)

ucimlrepo.dotdict.dotdict

Three Keys under the object; data: Contains dataset matrices as pandas dataframes; Contains metadata information about the dataset; Contains variable details presented in a tabular/dataframe format;

In [7]:
diabetes.keys()

dict_keys(['data', 'metadata', 'variables'])

under diabetes, "data"&"metadata" are dictionary,  "variables" is a dataframe

In [8]:
print(type(diabetes.data), type(diabetes.metadata), type(diabetes.variables))

<class 'ucimlrepo.dotdict.dotdict'> <class 'ucimlrepo.dotdict.dotdict'> <class 'pandas.core.frame.DataFrame'>


belows are the structure of "data","metadata"&"variables" 

In [9]:
print(diabetes["data"].keys())
print(diabetes['metadata'].keys())
print(diabetes['variables'])

dict_keys(['ids', 'features', 'targets', 'original', 'headers'])
dict_keys(['uci_id', 'name', 'repository_url', 'data_url', 'abstract', 'area', 'tasks', 'characteristics', 'num_instances', 'num_features', 'feature_types', 'demographics', 'target_col', 'index_col', 'has_missing_values', 'missing_values_symbol', 'year_of_dataset_creation', 'last_updated', 'dataset_doi', 'creators', 'intro_paper', 'additional_info'])
                        name     role         type demographic  \
0               encounter_id       ID                     None   
1                patient_nbr       ID                     None   
2                       race  Feature  Categorical        Race   
3                     gender  Feature  Categorical      Gender   
4                        age  Feature  Categorical         Age   
5                     weight  Feature  Categorical        None   
6          admission_type_id  Feature  Categorical        None   
7   discharge_disposition_id  Feature  Categorical    

store variable dataframe into dfvariables

In [10]:
dfvariables = diabetes["variables"]
dfvariables

name     role         type demographic  \
0               encounter_id       ID                     None   
1                patient_nbr       ID                     None   
2                       race  Feature  Categorical        Race   
3                     gender  Feature  Categorical      Gender   
4                        age  Feature  Categorical         Age   
5                     weight  Feature  Categorical        None   
6          admission_type_id  Feature  Categorical        None   
7   discharge_disposition_id  Feature  Categorical        None   
8        admission_source_id  Feature  Categorical        None   
9           time_in_hospital  Feature      Integer        None   
10                payer_code  Feature  Categorical        None   
11         medical_specialty  Feature  Categorical        None   
12        num_lab_procedures  Feature      Integer        None   
13            num_procedures  Feature      Integer        None   
14           num_medications  Feature      Integer        None   
15         number_outpatient  Feature      Integer        None   
16          number_emergency  Feature      Integer        None   
17          number_inpatient  Feature      Integer        None   
18                    diag_1  Feature  Categorical        None   
19                    diag_2  Feature  Categorical        None   
20                    diag_3  Feature  Categorical        None   
21          number_diagnoses  Feature      Integer        None   
22             max_glu_serum  Feature  Categorical        None   
23                 A1Cresult  Feature  Categorical        None   
24                 metformin  Feature  Categorical        None   
25               repaglinide  Feature  Categorical        None   
26               nateglinide  Feature  Categorical        None   
27            chlorpropamide  Feature  Categorical        None   
28               glimepiride  Feature  Categorical        None   
29             acetohexamide  Feature  Categorical        None   
30                 glipizide  Feature  Categorical        None   
31                 glyburide  Feature  Categorical        None   
32               tolbutamide  Feature  Categorical        None   
33              pioglitazone  Feature  Categorical        None   
34             rosiglitazone  Feature  Categorical        None   
35                  acarbose  Feature  Categorical        None   
36                  miglitol  Feature  Categorical        None   
37              troglitazone  Feature  Categorical        None   
38                tolazamide  Feature  Categorical        None   
39                   examide  Feature  Categorical        None   
40               citoglipton  Feature  Categorical        None   
41                   insulin  Feature  Categorical        None   
42       glyburide-metformin  Feature  Categorical        None   
43       glipizide-metformin  Feature  Categorical        None   
44  glimepiride-pioglitazone  Feature  Categorical        None   
45   metformin-rosiglitazone  Feature  Categorical        None   
46    metformin-pioglitazone  Feature  Categorical        None   
47                    change  Feature  Categorical        None   
48               diabetesMed  Feature  Categorical        None   
49                readmitted   Target  Categorical        None   

                                          description units missing_values  
0                   Unique identifier of an encounter  None             no  
1                      Unique identifier of a patient  None             no  
2   Values: Caucasian, Asian, African American, Hi...  None            yes  
3           Values: male, female, and unknown/invalid  None             no  
4   Grouped in 10-year intervals: [0, 10), [10, 20...  None             no  
5                                   Weight in pounds.  None            yes  
6   Integer identifier corresponding to 9 distinct...  None             no  
7   Integer identifier corresponding to 2

In [11]:
for desc in dfvariables[dfvariables["name"]=="readmitted"]["description"]:
    print(desc) #To get the description of a variable

Days to inpatient readmission. Values: <30 if the patient was readmitted in less than 30 days, >30 if the patient was readmitted in more than 30 days, and No for no record of readmission.


store the original data in to dforigin 

In [16]:
dforigin = diabetes.data.original
dforigin["medical_specialty"].unique() # Example of accessing the original data

array(['Pediatrics-Endocrinology', nan, 'InternalMedicine',
       'Family/GeneralPractice', 'Cardiology', 'Surgery-General',
       'Orthopedics', 'Gastroenterology',
       'Surgery-Cardiovascular/Thoracic', 'Nephrology',
       'Orthopedics-Reconstructive', 'Psychiatry', 'Emergency/Trauma',
       'Pulmonology', 'Surgery-Neuro',
       'Obsterics&Gynecology-GynecologicOnco', 'ObstetricsandGynecology',
       'Pediatrics', 'Hematology/Oncology', 'Otolaryngology',
       'Surgery-Colon&Rectal', 'Pediatrics-CriticalCare', 'Endocrinology',
       'Urology', 'Psychiatry-Child/Adolescent', 'Pediatrics-Pulmonology',
       'Neurology', 'Anesthesiology-Pediatric', 'Radiology',
       'Pediatrics-Hematology-Oncology', 'Psychology', 'Podiatry',
       'Gynecology', 'Oncology', 'Pediatrics-Neurology',
       'Surgery-Plastic', 'Surgery-Thoracic',
       'Surgery-PlasticwithinHeadandNeck', 'Ophthalmology',
       'Surgery-Pediatric', 'Pediatrics-EmergencyMedicine',
       'PhysicalMedicineandRe

dforigin breakdown

In [80]:
dforigin[dforigin["readmitted"] == ">30"]


encounter_id  patient_nbr             race  gender      age weight  \
1             149190     55629189        Caucasian  Female  [10-20)    NaN   
5              35754     82637451        Caucasian    Male  [50-60)    NaN   
7              63768    114882984        Caucasian    Male  [70-80)    NaN   
10             28236     89869032  AfricanAmerican  Female  [40-50)    NaN   
14             62256     49726791  AfricanAmerican  Female  [60-70)    NaN   
...              ...          ...              ...     ...      ...    ...   
101754     443842016    183087545        Caucasian  Female  [70-80)    NaN   
101755     443842022    188574944            Other  Female  [40-50)    NaN   
101756     443842070    140199494            Other  Female  [60-70)    NaN   
101760     443847176     50375628  AfricanAmerican  Female  [60-70)    NaN   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)    NaN   

        admission_type_id  discharge_disposition_id  admission_source_id  \
1                       1                         1                    7   
5                       2                         1                    2   
7                       1                         1                    7   
10                      1                         1                    7   
14                      3                         1                    2   
...                   ...                       ...                  ...   
101754                  1                         1                    7   
101755                  1                         1                    7   
101756                  1                         1                    7   
101760                  1                         1                    7   
101761                  1                         3                    7   

        time_in_hospital payer_code medical_specialty  num_lab_procedures  \
1                      3        NaN               NaN                  59   
5                      3        NaN               NaN                  31   
7                      5        NaN               NaN                  73   
10                     9        NaN               NaN                  47   
14                     1        NaN               NaN                  49   
...                  ...        ...               ...                 ...   
101754                 9        NaN               NaN                  50   
101755                14         MD               NaN                  73   
101756                 2         MD               NaN                  46   
101760                 6         DM               NaN                  45   
101761                 3         MC               NaN                  51   

        num_procedures  num_medications  number_outpatient  number_emergency  \
1                    0               18                  0                 0   
5                    6               16                  0                 0   
7                    0               12                  0                 0   
10                   2               17                  0                 0   
14                   5                2                  0                 0   
...                ...              ...                ...               ...   
101754               2               33                  0                 0   
101755               6               26                  0                 1   
101756               6               17                  1                 1   
101760               1               25                  3                 1   
101761               0               16                  0                 0   

        number_inpatient  diag_1  diag_2  diag_3  number_diagnoses  \
1                      0     276  250.01     255                 9   
5                      0     414     411     250                 9   
7                      0     428     492     250              

In [44]:
diabetes.data.features.columns

Index(['race', 'gender', 'age', 'weight', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed'],
      dtype='object')

In [58]:
diabetes.data.targets.columns

Index(['readmitted'], dtype='object')

In [76]:
dforigin.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)    NaN   
1        149190     55629189        Caucasian  Female  [10-20)    NaN   
2         64410     86047875  AfricanAmerican  Female  [20-30)    NaN   
3        500364     82442376        Caucasian    Male  [30-40)    NaN   
4         16680     42519267        Caucasian    Male  [40-50)    NaN   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital payer_code         medical_specialty  num_lab_procedures  \
0                 1        NaN  Pediatrics-Endocrinology                  41   
1                 3        NaN                       NaN                  59   
2                 2        NaN                       NaN                  11   
3                 2        NaN                       NaN                  44   
4                 1        NaN                       NaN                  51   

   num_procedures  num_medications  number_outpatient  number_emergency  \
0               0                1                  0                 0   
1               0               18                  0                 0   
2               5               13                  2                 0   
3               1               16                  0                 0   
4               0                8                  0                 0   

   number_inpatient  diag_1  diag_2 diag_3  number_diagnoses max_glu_serum  \
0                 0  250.83     NaN    NaN                 1           NaN   
1                 0     276  250.01    255                 9           NaN   
2                 1     648     250    V27                 6           NaN   
3                 0       8  250.43    403                 7           NaN   
4                 0     197     157    250                 5           NaN   

  A1Cresult metformin repaglinide nateglinide chlorpropamide glimepiride  \
0       NaN        No          No          No             No          No   
1       NaN        No          No          No             No          No   
2       NaN        No          No          No             No          No   
3       NaN        No          No          No             No          No   
4       NaN        No          No          No             No          No   

  acetohexamide glipizide glyburide tolbutamide pioglitazone rosiglitazone  \
0            No        No        No          No           No            No   
1            No        No        No          No           No            No   
2            No    Steady        No          No           No            No   
3            No        No        No          No           No            No   
4            No    Steady        No          No           No            No   

  acarbose miglitol troglitazone tolazamide examide citoglipton insulin  \
0       No       No           No         No      No          No      No   
1       No       No           No         No      No          No      Up   
2       No       No           No         No      No          No      No   
3       No       No           No         No      No          No      Up   
4       No       No           No         No      No          No  Steady   

  glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
0                  No                  No                       No   
1                  No                  No                       No   
2                  No                  No                       No   
3                  No                  No                       No   
4        